### DISCUSSION TITLE
Environment variables: boolean to enum conversion

### DISCUSSION THREAD

**Discussion by colinhacks** [geldata/gel#3049] [Pre-RFC]
Currently the EdgeDB CLI, `edgedb-docker`, and `edgedb-server` read several environment variables intended as boolean flags. Since environment variables are always strings, these flags need to be interpreted in some way. 

Options for boolean parsing:

1. Treat all non-empty values as `true`. This was deemed to susceptible to misconfiguration. A user may unknowingly change `EDGEDB_INSECURE_DEV_MODE=true` to `EDGEDB_INSECURE_DEV_MODE=false`, not realizing `"false"` is also considered truthy.
2. Support a set of boolean pairs (e.g. `yes/no`, `true/false`, `on/off`). When an unrecognized value is read, either a) throw an error or b) ignore + print a warning. It's odd and error-prone to have several ways to specify the same value.
3. Convert all boolean flags to enums. This opens the floor to future API expansion and isn't ambiguous.

It was decided to convert all boolean flags to enums. The goal of this discussion is to establish the new enum names and their allowable values.

### Context

As I was going through the options, few of the boolean flags lent themselves nicely to enum-ification. The insecure_dev_mode variable definitely should be an enum. I think allow_insecure_x_clients should as well. Everything else is much less clean cut.

There are a few flags that are unambiguously boolean (e.g. `EDGEDB_SERVER_GENERATE_SELF_SIGNED_CERT`) . Converting this into an enum with descriptive values seems like an unnecessary layer of complexity/indirection. Users will have to look up the valid values for each individual enum, and the enum-ified version of these variable names would be harder to understand.

So for all “boolean” variables I propose an enum consisting of two values: `true` and `false`. All environment variables should be named such that the default is false. It’s consistent and easy to remember.

Technically we could still support additional values in the future, e.g. `EDGEDB_SERVER_GENERATE_SELF_SIGNED_CERT=as_needed` (just a made-up example). This also lets us keep most variable names as-is, which decreases the surface area of the change. 

#### `EDGEDB_TLS_VERIFY_HOSTNAME`

New name: `EDGEDB_TLS_VERIFY_HOSTNAME`
Values: `false`, `true`
Default: `false`

#### `EDGEDB_NO_VERSION_CHECK` (CLI only)

⚠️ This should be renamed for consistency with other `SKIP_` variables.

New name: `EDGEDB_SKIP_VERSION_CHECK` 
Values: `true`, `false`
Default: `false`

#### `EDGEDB_SKIP_DOCKER_CHECK` (CLI only)

New name: `EDGEDB_SKIP_DOCKER_CHECK`
Values: `true`, `false`
Default: `false`

#### `EDGEDB_SERVER_ALLOW_INSECURE_BINARY_CLIENTS`

New name: `EDGEDB_SERVER_BINARY_SECURITY_MODE`
Values: `secure`, `insecure`
Default: `secure`

#### `EDGEDB_SERVER_ALLOW_INSECURE_HTTP_CLIENTS`

New name: `EDGEDB_SERVER_HTTP_SECURITY_MODE`
Values: `secure`, `insecure`
Default: `secure`

> Alternatively, we could combine these into `EDGEDB_SERVER_ALLOW_INSECURE_CLIENTS` which accepts a comma-separated list of `binary` or `http`. `""` = `[]`

#### `EDGEDB_SERVER_INSECURE_DEV_MODE`

New name: `EDGEDB_SERVER_SECURITY_MODE`
Values: `secure`, `insecure`
Default: `secure`

#### `EDGEDB_SERVER_SKIP_MIGRATIONS`

New name: *no change*
Values: `true`, `false`
Default: `false`

#### `EDGEDB_SERVER_GENERATE_SELF_SIGNED_CERT`

This could be implemented as a special value supplied to 

New name: no change
Values: `true`, `false`
Default: `false`

## edgedb.com

Unclear if either of these still do anything, but they're documented in various places.

#### `EDGEDB_DEBUG_HTTP_INJECT_CORS` 

New name: `EDGEDB_DEBUG_HTTP_CORS`
Values: `true`, `false`
Default: `true`

#### `EDGEDB_DEBUG_DISABLE_DOCS_EDGEQL_VALIDATION`

New name: no change
Values: `true`, `false`
Default: `false`

**Comment by tailhook** [comment]
My proposal:
1. `EDGEDB_TLS_VERIFY_HOSTNAME` -> **`EDGEDB_CLIENT_TLS_SECURITY`** (something along the lines of what Elvis proposed)
    * `insecure`
    * ~~`verify_certificate`~~ `no_host_verification`
    * ~~`verify_host`~~ `strict`
    * `default` -- is `strict` or `no_host_verification` depending on whether custom certificate is supplied
    * future extension `client_auth` (although can we have client authentication with or without verifying host?)
2. **`EDGEDB_CLIENT_SECURITY`**
    * `default`
    * `insecure_dev_mode`: for now, equivalent to setting `EDGEDB_CLIENT_TLS_SECURITY=insecure` though this mode may 
    encapsulate other behavior later
3. `EDGEDB_NO_VERSION_CHECK` -> **`EDGEDB_RUN_VERSION_CHECK`**
    * `never` -- skips the check
    * `default` == `cached`
    * possible extension `always` -- do not cache (mostly to suppress warning if no cache can be created)
    * possible extension `strict` -- fail if version cannot be checked or not the latest version is used
4. `EDGEDB_SKIP_DOCKER_CHECK` -> here are two options, because here it might be confusing with suppressing `--method=docker`, I appreciate feedback:
     * **`EDGEDB_INSTALL_IN_DOCKER`**
        * `forbid` == `default`
        * `allow` -- skips the check
        * possible extension `warn` -- do the check but print warning instead of forbidding
        * possible extension `strict` -- fail if could not check, currently warning is printed, but it probably means that CLI is running in a container with even more constraints, e.g. no access to `/proc` filesystem, so it's unlikely install will work)
    * ~~`EDGEDB_DOCKERIZED_MODE`~~
        * `no_install` or `no_instances` or `client_only` == `default`
        * `normal` -- allow installation
4. `EDGEDB_SERVER_ALLOW_INSECURE_BINARY_CLIENTS` -> **`EDGEDB_SERVER_BINARY_ENDPOINT_SECURITY`**
    * `tls` == `default`
    * `optional` -- allow no tls
    * possible extension `authenticated` or maybe `certificate` or maybe `client-cert` -- using client certificates
5. `EDGEDB_SERVER_ALLOW_INSECURE_HTTP_CLIENTS` -> **`EDGEDB_SERVER_HTTP_ENDPOINT_SECURITY`**
    * `tls` == `default`
    * `optional` -- allow no tls
    * possible extension `authenticated` or maybe `certificate` or maybe `client-cert` -- using client certificates
6. `EDGEDB_SERVER_INSECURE_DEV_MODE` -> ~~`EDGEDB_SERVER_PROFILE`~~ **`EDGEDB_SERVER_SECURITY`** / **`EDGEDB_CLIENT_SECURITY`**
    * `strict` == `default`
    * `insecure_dev_mode`
7. `EDGEDB_SERVER_SKIP_MIGRATIONS` -> **`EDGEDB_DOCKER_APPLY_MIGRATIONS`** (this is docker container feature)
    * `never`
    * `always` == `default` (or do we currently migrate on bootstrap only?)
    * possible extension: `on_bootstrap`
    * possible extension: `watch` -- watch changed files and apply migrations automatically
8. `EDGEDB_SERVER_GENERATE_SELF_SIGNED_CERT` -> **`EDGEDB_SERVER_TLS_CERT_MODE`** (or `ACQUIRE_CERT`?)
    * ~~`ensure_self_signed` or~~ `generate_self_signed` -- generate self-signed on bootstrap and when expired
    * ~~`required` or `passive` or `provided`~~ `require_file` == `default`
    * possible extension `letsencrypt`
    * possible extension `acme` or even `acme://corporate.verifier.example.org`
9. `EDGEDB_HIDE_GENERATED_CERT` -> **`EDGEDB_DOCKER_SHOW_GENERATED_CERT`**
    * `always` == `default`
    * `never`
  
Notes:
* `x or y` means we have to choose one option to implement
*  `==` means alias, both work (currently use it for `default` only)
* `default` is always an alias to whatever default is, as well as empty string (and non existing value)

Site/debug variables, like various log things in server could stay with whatever method is easier to implement. For python it might be click's validation for python (although for debug flags in server we search for emptyness in some places and for `not in {'',  '0`}' in others, which is also fine).

**Reply by elprans** [reply]
Overall, I like this.

> `EDGEDB_SERVER_GENERATE_SELF_SIGNED_CERT` -> **`EDGEDB_SERVER_CERT_MODE`** (or `ACQUIRE_CERT`?)

I suggest **`EDGEDB_SERVER_TLS_CERT_ISSUER`**: `self`, `user`, `letsencrypt`, ...

> `EDGEDB_SERVER_INSECURE_DEV_MODE` -> **`EDGEDB_SERVER_PROFILE`**

I suggest **`EDGEDB_SERVER_SECURITY`** to match the other `SECURITY` flags.

> `EDGEDB_SKIP_DOCKER_CHECK`

Perhaps have a higher-level `EDGEDB_INSTALLATION_MODES=native,docker,...`?  Could also set the method priority.

**Reply by 1st1** [reply]
Yeah, +1 for this proposal. I like that it avoids true/false values (that can still be confusing and one might pass "on" thinking it's the same as passing "true"). It's also extensible.

**Reply by tailhook** [reply]
> EDGEDB_SERVER_TLS_CERT_ISSUER=self

This doesn't say anything to me. There should be something saying `self-signed`.

`user` is also doesn't say anything. Who is that user? What it cert is from real ssl provider? If it's certificate issued by `letsencrypt` but provided as a file, should you configure `issuer=letsencrypt`?



> > `EDGEDB_SKIP_DOCKER_CHECK`
> 
> Perhaps have a higher-level `EDGEDB_INSTALLATION_MODES=native,docker,...`? Could also set the method priority.

This check prevents from installing server when edgedb is running **inside** the container. It has nothing to do to installation modes(methods).

**`EDGEDB_SERVER_SECURITY`** looks good.

**Reply by elprans** [reply]
> This doesn't say anything to me

OK, let's go with `EDGEDB_SERVER_TLS_CERT_MODE` (your original suggestion + `_TLS` for consistency).

> This check prevents from installing server when edgedb is running inside the container

Oh, OK, then `EDGEDB_INSTALL_IN_DOCKER` is probably an OK name.  It's a fairly marginal setting, so I wouldn't fret too much about it.

**Reply by colinhacks** [reply]
Small recommendation. I think it's more clear if variables that affect whether or not a given action is performed contain the action verb in the name. This would affect these three:

`EDGEDB_CHECK_VERSION` -> `EDGEDB_RUN_VERSION_CHECK`

`EDGEDB_CHECK_VERSION` is particularly confusing, it seems like the user should be specifying a version of something.

`EDGEDB_DOCKER_MIGRATIONS` -> `EDGEDB_DOCKER_APPLY_MIGRATIONS` 

`EDGEDB_DOCKER_GENERATED_CERT` -> `EDGEDB_DOCKER_SHOW_GENERATED_CERT`

**Reply by elprans** [reply]
> `EDGEDB_SERVER_BINARY_CLIENT_SECURITY`
> `EDGEDB_SERVER_HTTP_CLIENT_SECURITY`

I propose `EDGEDB_SERVER_BINARY_ENDPOINT_SECURITY` and `EDGEDB_SERVER_HTTP_ENDPOINT_SECURITY` to avoid having both `CLIENT` and `SERVER` in the name.
